### PR TITLE
Add library documentation

### PR DESCRIPTION
**Pull request by piercefreeman**
In order to encourage wider use and understanding of the microcosm dependency injection framework, let's add more easily accessible documentation that tells a story about the opinionated architecture choices.  This PR introduces a first pass at sphinx documentation generation, which creates a browsable index of help pages and API protocols.

This PR is a start but certainly not complete.  To get us started:

- Move the version number inside the library instead of locking within `setup.py`.  The principal motivation here was referencing the version number in the docs by importing the latest signature from the library during build-time.  Even outside of document generation, though, this is widely accepted practice so callers are able to easily inspect microcosm for the current version.
- Added basic documentation boilerplate: README references, CONTRIBUTING, etc.
- Switched primary documents to `rst` instead of markdown.  Sphinx support for markdown is lacking, so converting the entire document hierarchy reduces some coordination overhead and risk for formatting issues.  Writing rst has a different syntax than what we're used to, but I've found it very interpretable.  Making the leap shouldn't introduce much additional strain on documentation writers.
- Added basic API documentation.  As a POC, I added auto doc generation to `ObjectGraph` and `create_object_graph` since these are the main entrypoints to microcosm.  This documentation is generated dynamically from the docstring that's stored in the code itself.  It's still on us to customize exactly which classes and functions we choose to advertise within the API documentation.

Next Steps
- Make the API Documentation more robust by adding additional references / docstrings for common classes and functions
- CircleCI support for publishing documents on every release

### CODE REVIEW

**Comment by piercefreeman** [comment]
Note - this is going to fail CI until we agree on accepting `RST` as a valid documentation format and update `globality-build`.  The current harness generates a Dockerfile that copies over `README.md` - since that no longer exists, the pipeline fails early.

File: docs/installation.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+From sources
+------------
+
+The sources for Microcosm can be downloaded from the `Github repo`_.
+
+You can either clone the public repository:
+
+.. code-block:: console
+
+    $ git clone git://github.com/piercefreeman/microcosm
```
Comment (pokey) [inline docs/installation.rst]:
```suggestion
    $ git clone git://github.com/globality-corp/microcosm
```

File: docs/installation.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Or download the `tarball`_:
+
+.. code-block:: console
+
+    $ curl -OJL https://github.com/piercefreeman/microcosm/tarball/master
```
Comment (pokey) [inline docs/installation.rst]:
```suggestion
    $ curl -OJL https://github.com/globality-corp/microcosm/tarball/master
```

File: docs/installation.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+Once you have a copy of the source, you can install it with:
+
+.. code-block:: console
+
+    $ python setup.py install
+
+
+.. _Github repo: https://github.com/piercefreeman/microcosm
```
Comment (pokey) [inline docs/installation.rst]:
```suggestion
.. _Github repo: https://github.com/globality-corp/microcosm
```

File: docs/installation.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+.. _tarball: https://github.com/piercefreeman/microcosm/tarball/master
```
Comment (pokey) [inline docs/installation.rst]:
```suggestion
.. _tarball: https://github.com/globality-corp/microcosm/tarball/master
```

File: docs/Makefile [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
```
Comment (GabeLambda) [inline docs/Makefile]: Sphinx?

Comment (GabeLambda) [reply docs/Makefile]: Did you mean to check this file in?

File: docs/conf.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+# -- General configuration ---------------------------------------------
+
+# If your documentation needs a minimal Sphinx version, state it here.
+#
+# needs_sphinx = '1.0'
+
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
+extensions = ['sphinx.ext.autodoc', 'sphinx.ext.viewcode']
```
Comment (GabeLambda) [inline docs/conf.py]: Sphinx?

File: docs/Makefile (hunk 5 above)
Comment (piercefreeman) [reply docs/Makefile]: http://www.sphinx-doc.org/en/master/

File: docs/make.bat [hunk 7]
```diff
@@ -0,0 +1,36 @@
+@ECHO OFF
+
+pushd %~dp0
+
+REM Command file for Sphinx documentation
```
Comment (GabeLambda) [inline docs/make.bat]: Sphinx?

**Comment by piercefreeman** [comment]
@pokey Good question. The docs checked in to `docs` are just the template files; in order to be rendered into the full documentation format you need to run the sphinx generation process.  Something like:

`cd docs && make singlehtml`

File: docs/Makefile (hunk 5 above)
Comment (GabeLambda) [reply docs/Makefile]: Ha! I completely forgot about this project (which is sad, because I've used it in the past...)

File: .bumpversion.cfg [hunk 8]
```diff
@@ -1,9 +1,9 @@
 [bumpversion]
-current_version = 2.13.1
+current_version = 2.13.2
```
Comment (cdm-ium) [inline .bumpversion.cfg]: I'd probably bump minor version here, not patch, this isn't a bug fix.

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+   :alt: Circle CI
+
+
+Terminology
+-----------
+
+
+* A ``microservice`` is a small software application. It is composed of several smaller pieces
+  of software, many of which are reusable.
+* A ``component`` is one of these (possibly reusable) pieces of software.
```
Comment (cdm-ium) [inline README.rst]: Definitely reusable!

File: README.rst [hunk 10]
```diff
@@ -0,0 +1,144 @@
+
+Simple Microservice Configuration
```
Comment (cdm-ium) [inline README.rst]: I love this write up, just not sure it justifies the wholesame conversion to rst. This library doesn't change a lot and I don't forsee the need to add significantly more docs.

File: docs/installation.rst [hunk 11]
```diff
@@ -0,0 +1,51 @@
+.. highlight:: shell
+
+============
```
Comment (cdm-ium) [inline docs/installation.rst]: Realistically this is a lot of writing (e.g. overhead) for the main point that "microcosm is available on pypi".

File: docs/usage.rst [hunk 12]
```diff
@@ -0,0 +1,7 @@
+=====
```
Comment (cdm-ium) [inline docs/usage.rst]: Is this unfinished? At least from the file name, the point of this file seems superceded by the contents of README.rst.

**Comment by jameschung2019** [comment]
Hey there, please close this PR or merge it. It will be auto-closed in 48 hours if not acted upon. If you decide to merge it, comment and take ownership to coordinate, schedule, and get it done. Thanks!

**Comment by jameschung2019** [comment]
This PR is being closed per CTO leadership's decision.  All PRs older than 1/1/2020 is included.

**Pull request closed without merging**